crates/pathfinder/src/state/contract_hash.rs

```diff
@@ -0,0 +1,413 @@
+    use json::EntryPointType::*;
+    // related python impl: https://github.com/starkware-libs/cairo-lang/blob/7712b21fc3b1cb02321a58d0c0579f5370147a8b/src/starkware/starknet/core/os/contract_hash.py
+
+    let mut contract_definition =
+        serde_json::from_str::<json::ContractDefinition>(contract_definition_dump)
+            .context("Parse contract_definition")?;
+
+    // the other modification is handled by skipping if the attributes vec is empty
+    contract_definition.program.debug_info = None;
+
+    let truncated_keccak = {
+        let mut ser =
+            serde_json::Serializer::with_formatter(KeccakWriter::default(), PythonDefaultFormatter);
+
+        contract_definition
+            .serialize(&mut ser)
+            .context("Serializing contract_definition for Keccak256")?;
+
+        let KeccakWriter(hash) = ser.into_inner();
+        truncated_keccak(<[u8; 32]>::from(hash.finalize()))
+    };
+
+    // what follows should be up to date with https://github.com/starkware-libs/cairo-lang/blob/64a7f6aed9757d3d8d6c28bd972df73272b0cb0a/src/starkware/starknet/core/os/contracts.cairo#L76-L118
+
+    const API_VERSION: StarkHash = StarkHash::ZERO;
+
+    let mut outer = HashChain::default();
+
+    outer.update(API_VERSION);
+
+    // It is important process the different entrypoint hashchains in correct order.
+    // Each of the entrypoint lists gets updated into the `outer` hashchain.
+    [External, L1Handler, Constructor]
+        .iter()
+        .map(|key| {
+            contract_definition
+                .entry_points_by_type
+                .get(key)
+                .unwrap_or(&Vec::new())
+                .iter()
+                .enumerate()
+                // flatten each entry point to get a list of (selector, offset, selector, offset, ...)
+                // `i` is the nth selector of the `key` kind
+                .flat_map(|(i, x)| {
+                    [("selector", &*x.selector), ("offset", &*x.offset)]
+                        .into_iter()
+                        .map(move |(field, x)| match x.strip_prefix("0x") {
+                            Some(x) => Ok((field, x)),
+                            None => Err(anyhow::anyhow!(
+                                "Entry point missing '0x' prefix under {key} at index {i} entry ({field})",
+                            )),
+                        })
+                        .map(move |res| {
+                            res.and_then(|(field, hex)| {
+                                StarkHash::from_hex_str(hex).with_context(|| {
+                                    format!("Entry point invalid hex under {key} at index {i} entry ({field})")
+                                })
+                            })
+                        })
+                        .into_iter()
+                })
+                .try_fold(HashChain::default(), |mut hc, next| {
+                    hc.update(next?);
+                    Result::<_, Error>::Ok(hc)
```

Comment:
These hints are needed when working with try_fold, try_for_each. Otherwise there seems to be a problem determining the Err type from the `?`.

Reply:
Unified these a bit.